controller/controller.go

```diff
@@ -248,75 +246,85 @@ func (c *ObjectStorageController) runWorker(ctx context.Context) {
 
 func (c *ObjectStorageController) processNextItem(ctx context.Context) bool {
 	// Wait until there is a new item in the working queue
-	op, quit := c.queue.Get()
+	uuidInterface, quit := c.queue.Get()
 	if quit {
 		return false
 	}
 
+	uuid := uuidInterface.(types.UID)
+	var err error
 	// With the lock below in place, we can safely tell the queue that we are done
 	// processing this item. The lock will ensure that multiple items of the same
 	// name and kind do not get processed simultaneously
-	defer c.queue.Done(op)
+	defer c.queue.Done(uuid)
 
 	// Ensure that multiple operations on different versions of the same object
 	// do not happen in parallel
-	c.OpLock(op)
-	defer c.OpUnlock(op)
+	c.OpLock(uuid)
```

Comment:
Sorry, I am missing something here. How does tracking uuid prevent operating in parallel on different _versions_of the same object? Wouldn't this serialize on all version of the object?

Reply:
In our controller, we run a single shared informer between multiple crd resource event handlers. The single shared informer reports changes in bucket* objects as they occur. The events are processed by first encoding them in operation structs, and putting them into the tail of a workqueue. The workqueue implementation is what this PR mostly touches on. 

The workqueue was previously holding operation objects (addOp, updateOp, deleteOp). On any event, these operation objects were put into the queue, and waited to be processed by the next available thread.

These objects were not comparable with one another, and that lead to a few problems:

 -  Gven an object, if update occurs while add is waiting in the queue, it should only result in 1 event - add, with the updated fields
 -  Gven an object, if update occurs while another update is waiting in the queue, then the second update is the only one that should be processed
 
The previous implementation, albeit technically eventually consistent, could not ensure the above dynamics. This was simply because it was not possible to determine new events on an object already present in the queue. 

Now, for the` uuid`. This is a string field that is easily comparable (`==`), and by using the kubernetes UID as the unique value in this field, it is possible to compare two events occurring on the same object. 

This change also allowed me to improve the exponential backoff characteristics. If new events occur while an operation on the same object is in back-off, the new event will not be processed before the back-off period is complete. i.e. the uuid positionin the queue is not changed. Instead, the value of `opMap[uuid]` is set to the new operation structure denoting the event. By decoupling the operation itself from the queuing order, we gain a lot of flexibility.

Now to answer your direct question,

> Sorry, I am missing something here. How does tracking uuid prevent operating in parallel on different versions of the same object? Wouldn't this serialize on all versions of the object?

By using uuid, we do not serialize the steps. We either replace old events with new ones if both are updates or handle them differently based on the type of the events.